### ISSUE TITLE
[css-pseudo-4] Custom state pseudo class proposal

### ISSUE THREAD

**Issue by atanassov** [w3c/csswg-drafts#4805]
Based on recent [TAG review](https://github.com/w3ctag/design-reviews/issues/428) of the proposed [Custom State Pseudo Class](https://wicg.github.io/custom-state-pseudo-class) feature it was surprising that this topic hasn't been discussed with the csswg yet.

Given Chrome has an [intent to ship](https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/dJibhmzE73o) this feature it will be helpful for us to at least discuss it.

**Comment by SelenIT** [comment]
Given that this proposal seems to be custom-element-specific, probably [css-shadow-parts-1] would be a better place for it than [css-pseudo-4]?

**Comment by heycam** [comment]
> it was surprising that this topic hasn't been discussed with the csswg yet

Procedurally, is it incumbent on us as CSSWG members to notice specifications developed in WICG that impact CSS, and to then bring them up for discussion here to provide feedback in that venue?

**Comment by tabatkins** [comment]
No, it's supposed to be part of our process to rope in appropriate experts, and in this case would absolutely be a "file an issue in CSSWG" sort of thing. Failing on our part, sorry about that; I didn't realize we were at the ItS stage yet.

**Comment by smfr** [comment]
How does this relate to https://github.com/w3c/csswg-drafts/issues/3431?

**Comment by FremyCompany** [comment]
(subscribing myself to this issue)

**Comment by fantasai** [comment]
CSSWG minutes are here: https://lists.w3.org/Archives/Public/www-style/2020Mar/0009.html
An important comment was to make sure this feature can handle enums, not just boolean states.

**Comment by tabatkins** [comment]
Web Components f2f minutes here: https://github.com/WICG/construct-stylesheets/issues/45#issuecomment-602756608

Overall there's consensus on implementing `document.adoptedStyleSheets` as an `ObservableArray`.

There also appears to be consensus on moving `document.styleSheets` to a (readonly) `ObservableArray` subclass, so it'll sprout the same array methods.

The contentious question that couldn't be resolved on the call, and which is being left for us to decide, is whether direct assignment of an array is allowed; that is, whether `document.aSS = [ss1, ss2];` works or throws.

Points in favor:
* Other mutating list-like APIs on the platform allow direct assignment, or plan to: `el.classList` allows it; Typed OM array-likes intend to allow it; various CSSOM APIs (`CSSMediaRule.media` and `CSSStyleRule.style` effectively allow it, via their `[PutForwards]` extended attributes that allow setting to a string to replace the associated lists).
* Seems weird to disallow direct setting, when there are other ways to replace the entire array (calling `.splice()`, or emptying the array and then `.push()`ing). Complex restrictions seem bad without strong reasons for them, as it means authors have to learn a more complex API boundary and can't rely on knowledge from elsewhere.

Points against:
* When subclassing a custom element, setting an array assumes that the superclass hasn't set any stylesheets itself; if it has, and you weren't aware of this, the subclass setting an array will wipe out the superclass's sheets. If the subclass instead just always used `.push()`, it wouldn't clobber anything.
* The setter will consume the array, not use it; `document.aSS` will have equal *contents* to the array you set, but not equal *identity* (it'll return an ObservableArray object, not the iterable the author assigned). This is distinct from how userland array-valued properties tend to work, and might be weird.

----

I'm at this point strongly on the "assignment should be allowed" camp. In particular, the first point against was most vociferously argued, and I think it's absolutely something we should reject.  The footgun **is not** "clobbering the superclass's styles", it's subclassing at all. It is exactly as dangerous to naively clobber the superclass's styles as it is to naively append to the superclass's (unknown) styles; both are equally likely to screw up your component with parts not getting the styles they expected.

In addition, this problem is common to *every single aspect of the superclass*, not just styles. Literally *anything* the subclass does might, if it's not aware of what the superclass is doing, screw up what the superclass is doing. Adding a special-case restriction to try and make subclasses less likely to screw up *this specific case* won't materially improve the situation. Even if it *weren't* the case that I think `.push()`ing is just as dangerous (so there's *no way* to reduce the risk anyway), increasing the complexity of the overall API surface for an insignificant improvement in safety is generally a bad tradeoff, I believe. We've made the judgement that similar sorts of tradeoffs aren't worthwhile elsewhere in CSS.

------

So anyway, Agenda+ to get a final decision on this. I secured a promise from Apple to ensure a relevant person attends the CSSWG call, so hopefully this can be bumped for this week's agenda?

**Comment by othermaciej** [comment]
Constructible Stylesheets seems like a separate issue and separate spec from Custom State Pseudo Class, perhaps it merits a separate issue?

**Comment by tabatkins** [comment]
Whoops, yeah, I posted this to the wrong place. I'll move it appropriately.

**Comment by josepharhar** [comment]
> An important comment was to make sure this feature can handle enums, not just boolean states.

I found a bunch of discussion including what looks like a concrete proposal here: https://github.com/WICG/custom-state-pseudo-class/issues/4#issuecomment-632556585
What does everyone think?

**Comment by josepharhar** [comment]
I started a PR: https://github.com/w3c/csswg-drafts/pull/8213

**Comment by emilio** [comment]
Agenda+ as per @josepharhar's request, re. the discussion about `:state(foo)` and `:--foo`

**Comment by emilio** [comment]
Some of that discussion is in https://github.com/whatwg/html/pull/8467, ftr.

**Comment by josepharhar** [comment]
context: I don't have an opinion about `:state(foo)` vs `:--foo`, but @rniwa thought that we should be doing `:state` here: https://github.com/whatwg/html/pull/8467#issuecomment-1648866722
There was also a previous discussion/resolution in the CSSWG here: https://lists.w3.org/Archives/Public/www-style/2020May/0017.html

**Comment by Loirooriol** [comment]
#8094 points out that `:--foo` is also proposed for custom selectors, e.g. https://drafts.csswg.org/css-extensions/#custom-selectors has this example:

```css
@custom-selector :--heading {
  expansion: h1, h2, h3, h4, h5, h6;
}
```

**Comment by josepharhar** [comment]
The `:--foo` syntax has been shipping in stable chrome for some time now, with a usecounter at 0.03%: https://chromestatus.com/metrics/feature/timeline/popularity/3796
I'm going to send an intent to deprecate for the old syntax with a plan to replace it with the new syntax, but I am concerned that my proposal to change the syntax may be rejected. I will post updates here.
@rniwa @annevk

**Comment by annevk** [comment]
The HTML side of this has merged. But it also needs to be added to Selectors for parsing purposes. Semantics can be left to the host.

**Comment by ShobujDash** [comment]
good

**Comment by mangelozzi** [comment]
Chrome warns one with a deprecation warning like this:
> [Deprecation] Custom state pseudo classes have been changed from ":--spacer" to ":state(spacer)". See more here: https://github.com/w3c/csswg-drafts/issues/4805. 

It points one to this issue, which I think is an odd place to send one for help on an issue. This issue just seems to be a list of people posting in the wrong place and moving threads with no explaination about the deprecation warnging. 

I don't understand what I am supposed to change, this is the code it flags for the warning above:
```js
    getIsAboveBreakpoint() {
        const breakpoint = this.getAttribute('breakpoint');
        return this.elToWatch.clientWidth >= breakpoint;
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^
    }
```

**Comment by Loirooriol** [comment]
This issue contains the resolution that changed the syntax: https://github.com/w3c/csswg-drafts/issues/4805#issuecomment-1663111980

Complain to Chrome about pointing to a better resource and blaming the wrong line of code (that code isn't even CSS). This is not the place to have a tutorial.

**Comment by MaherbanaliJ** [comment]
[Deprecation] Custom state pseudo classes have been changed from ":--webkit-search-cancel-button" to ":state(webkit-search-cancel-button)"

in my case this error show in my medusa js storefront.

**Comment by josepharhar** [comment]
I am removing the console message in chrome 133, sorry for the noise
